src/libraries/System.Private.CoreLib/src/System/Half.cs

```diff
@@ -0,0 +1,706 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Globalization;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System
+{
+    // Portions of the code implemented below are based on the 'Berkeley SoftFloat Release 3e' algorithms.
+
+    /// <summary>
+    /// An IEEE 754 compliant float16 type.
+    /// </summary>
+    [Serializable]
+    [StructLayout(LayoutKind.Sequential)]
+    public readonly struct Half : IComparable, IFormattable, IComparable<Half>, IEquatable<Half>, ISpanFormattable
```

Comment:
@TamarChristinaArm, I was llooking at the ARM calling convention for `half-precision floating point` and it looks like its treated as a `ushort` rather than as a floating-point type for the purposes of argument passing/returning, is that correct?

> A Half-precision Floating Point Type is returned in the least significant 16 bits of r0.
> If the argument is a Half-precision Floating Point Type its size is set to 4 bytes as if it had been copied to the least significant bits of a 32-bit register and the remaining bits filled with unspecified values.

There also seems to be a section on variants of the calling convention which details:
> A half precision floating point type is passed as if it were loaded from its memory format into the least significant 16 bits of a single precision register.

Reply:
>  its treated as a ushort rather than as a floating-point type for the purposes of argument passing/returning, is that correct?

@tannergooding passing it in `GPR` is only for the `base` variant, i.e. `soft` or `softfp` float ABI where you don't have hardware support for floats. When you have hardware support and use the `hard` float ABI it uses the `VFP` calling convention which passes it in the `s` register as the second part you quoted.

I'm guessing coreclr only supports `hard`? in which case the `VFP` one is what you want :).

Reply:
@tannergooding we need to know the type when we call and when we return, and without retyping all logic is located in `Compiler::getReturnTypeForStruct`, so necessary changes would be localized in the JIT inside this function and functions that are called by it. 

I think it could be done similar to #37499 by adding a new type into VM and saying that we are returning it as a float it JIT, lower will add bit casts as necessary. 

Do you expect methods that are using `private readonly ushort m_value;` directly, like:
```
        public static bool IsPositiveInfinity(Half value)
        {
            return value.m_value == PositiveInfinityBits;
        }
```
would get `value` as `float`, but access its field as a short that would case a memory spill?

Reply:
Thanks for the explanation!

> would get value as float, but access its field as a short that would cause a memory spill?

We have similar logic for `System.Single` and `Siystem.Double`: <https://source.dot.net/#System.Private.CoreLib/Single.cs,79>
Prior to .NET 5 it did cause a memory spill but I believe it now just emits a `movd` to directly transition between the floating-point and integer registers.

https://github.com/dotnet/runtime/issues/11413 tracks the more general issue there and is something we could workaround for the type on ARM32/ARM64 as well.
I don't believe it will be a problem for x86/x64 as none of the architectures we support define an ABI for `Half` and so passing it around as an integer is as expected.

Reply:
So ... I could probably look this up, but I'm guessing you already know, so I'll ask here: Is `Half` considered a floating point type for the purposes of HFA recognition? If so, I think there might be a bit more work in the JIT to handle non-singletons correctly.

Reply:
The short answer is yes. The actual handling of ARM32 vs ARM64 appears to differ slightly and the docs on ARM64 appear to be quite a bit clearer.

The ARM64 specs are apparently on GitHub now 😄: https://github.com/ARM-software/abi-aa/releases
See also: https://developer.arm.com/architectures/system-architectures/software-standards/abi